### PR TITLE
OpenAI inference params

### PR DESCRIPTION
**Pull request by santiagxf**
[Open in GitHub Codespaces](https://codespaces.new/santiagxf/mlflow/pull/9909?quickstart=1)

### Related Issues/PRs

Resolve:

* https://github.com/mlflow/mlflow/issues/9678

### What changes are proposed in this pull request?

Introducing support for inference parameters in OpenAI flavor in MLflow.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introducing support for inference parameters in the predict method for OpenAI flavor.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-openai-params"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by santiagxf** [comment]
@BenWilson2 @harupy, this is the last PR for all the improvements for OpenAI flavor :D

File: mlflow/utils/class_utils.py [hunk 1]
```diff
@@ -1,6 +1,19 @@
...
 import importlib
+import inspect
 
 
 def _get_class_from_string(fully_qualified_class_name):
     module, class_name = fully_qualified_class_name.rsplit(".", maxsplit=1)
     return getattr(importlib.import_module(module), class_name)
+
+
+def nameof(_) -> str:
```
Comment (BenWilson2) [inline mlflow/utils/class_utils.py:11]: This is exceptionally clever. 

However, do we really want to be disassembling bytecode to get the name of a variable? 

for the current use case that this is being used for, is there a compelling downside to specifying the name of that variable as a constant?

p.s., I did grin when I saw this.

Comment (harupy) [reply mlflow/utils/class_utils.py:11]: I'm not a fan of a trick like this. Can we just hardcode?

Comment (santiagxf) [reply mlflow/utils/class_utils.py:11]: I completely see your point. Let's hardcode `batch_name` and revisit it when this is a frequent enough problem.

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks @santiagxf :)

File: mlflow/openai/utils.py [hunk 2]
```diff
@@ -135,6 +135,20 @@ def request(*args, **kwargs):
     return _mock_request(new=request)
 
 
+def _validate_model_params(task, model, params):
+    if params:
+        if any(param for param in params.keys() if param in model.keys()):
```
Comment (harupy) [inline mlflow/openai/utils.py:140]:
```suggestion
        if any(key for key in params if key in model):
```

File: mlflow/openai/__init__.py [hunk 3]
```diff
@@ -346,11 +346,16 @@ def save_model(
     _validate_and_prepare_target_save_path(path)
     code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
     task = _get_task_name(task)
+    _get_api_config()
```
Comment (harupy) [inline mlflow/openai/__init__.py:349]: @santiagxf Why do we need to call `_get_api_config` here?

Comment (harupy) [reply mlflow/openai/__init__.py:349]: > @santiagxf Why do we need to call `_get_api_config` here?

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/9909#discussion_r1360044714 is addressed!

File: mlflow/openai/__init__.py (hunk 3 above)
Comment (santiagxf) [reply mlflow/openai/__init__.py]: Great catch. I must have had it on the clipboard and pasted it accidentally. Indeed it makes no sense.

**Comment by harupy** [comment]
@santiagxf can you rebase on master branch to resolve the conflict?

**Pull request merged**